### DISCUSSION TITLE
How to "intelligently" compress an image in its original format by auto choosing encoder and making sure the compression makes sense?

### DISCUSSION THREAD

**Discussion by felinepc** [SixLabors/ImageSharp#1651] [Q&A]
Currently, I use ImageSharp to resize images and save them in their original formats, via the codes below:

```
(Image Image, IImageFormat Format) imgWithFormat = await Image.LoadWithFormatAsync(stream);

using var image = imgWithFormat.Image;

image.Mutate(x => x.Resize(
        new ResizeOptions() 
        {
            Mode=ResizeMode.Max,
            Size = new Size(1280, 1280)
        }));

using var ms = new MemoryStream();

await image.SaveAsync(ms, imgWithFormat.Format);

ms.Position = 0;

//... ...
```

Here I resized the image first if it's too big via `ResizeMode.Max`.

What I want to do next, is to optimize the image similar to the fashion of [https://tinypng.com/](https://tinypng.com/), where:

1. The original format is always kept, i.e. jpg/png/gif etc. It seems ImageSharp expects me to pass an specific encoder when encoding.

2. The compression level should be of a quality I specify (BTW, I can find quality property for `JpegEncoder` and `PngEncoder`, but not `GifEncoder`). Is it okay to leave every other encoder option on default?

3. If the original image is already compressed to around the ideal level, the re-encoding should not result in significant quality loss or file size increase. This is what I observed from services like tinypng. Better yet, is it even possible to detect this "ideal size" using some sort of metrics and not re-encode?

Your advice would be greatly appreciated!

**Comment by JimBobSquarePants** [comment]
>1. The original format is always kept, i.e. jpg/png/gif etc. It seems ImageSharp expects me to pass an specific encoder when encoding.

That's not true. The library will always use the input format unless specified otherwise.

>2. The compression level should be of a quality I specify (BTW, I can find quality property for JpegEncoder and PngEncoder, but not GifEncoder). Is it okay to leave every other encoder option on default?

There is no `Quality` property in the `PngEncoder`. Quality is a property of the `JpegEncoder`. The concept simply doesn't exist with other codecs. There are plenty of documented configuration options for [each](https://docs.sixlabors.com/api/ImageSharp/SixLabors.ImageSharp.Formats.Png.PngEncoder.html#properties) of [those](https://docs.sixlabors.com/api/ImageSharp/SixLabors.ImageSharp.Formats.Gif.GifEncoder.html#properties).

>3. If the original image is already compressed to around the ideal level, the re-encoding should not result in significant quality loss or file size increase. This is what I observed from services like tinypng. Better yet, is it even possible to detect this "ideal size" using some sort of metrics and not re-encode?

Why are you comparing specialized optimization services like tinypng?

ImageSharp detects many metadata properties and will attempt to encode output images using the same configuration. In some instances this is lossy and expected (jpeg) and you might be better specifying a higher quality to encode at to preserve more of the detail.

There's no mechanism for detecting best options. What you are asking for is machine learning.

**Reply by felinepc** [reply]
@JimBobSquarePants Thanks a lot for the clarifications. What I meant about "specific encoder" is the encoder option part: because I do not know what kind of image the user uploaded yet, I do not know what kind of `IImageEncoder` to pass to the `SaveAsync` method.  I realized I could use a `switch` statement on the returned `ImageFormat` to return the correct encoder with my preferred options, just want to make sure if this the right way.

Sorry I did not realize a service like tinypng has a lot more going under the hood for optimization. I just assumed `ImageSharp` had similar capabilities built in. First time working with this stuff!

Based on what you told me about `ImageSharp`'s meta data detection already attempting to use a sensible configuration, I think I'd just resize it with my current code as is, and let `ImageSharp` choose the encoder options as it sees fit when saving. This would probably work just fine for me.
